Fabric.Authorization.API/scripts/AzureAD-Migration.ps1

```diff
@@ -0,0 +1,252 @@
+                    $authenticationFailed = $true
+                }
+                else {
+                    Write-DosMessage -Level Information -Message "Unexpected error when retrieving Azure AD group. $($_.Exception)"
+                    throw
+                }
+            }
+            catch {
+                $errorMsg = "Unexpected error while connecting to Azure AD: $($_.Exception)"
+                Write-DosMessage -Level Error -Message $errorMsg
+                throw $errorMsg
+            }
+        } while ($authenticationFailed -eq $true)
+    }
+
+    # disconnect from Azure AD
+    try {
+        Disconnect-AzureAD
+    }
+    catch {
+        $errorMsg = "There was an unexpected error error while disconnecting from Azure AD tenant: $($tenantId)"
+        Write-DosMessage -Level Error -Message $errorMsg
+        throw $errorMsg
+    }
+
+    if ($azureADGroups.Count -eq 1) {
+        return $azureADGroups[0]
+    }
+    elseif ($azureADGroups.Count -eq 0) {
+        $errorMsg = "No match found for SID $($groupSID) in Azure AD Tenant $($tenantId)."
+        Write-DosMessage -Level Information -Message $errorMsg
+        throw $errorMsg
+    }
+    else {
+        $errorMsg = "Multiple matches found for group SID $($groupSID)."
+        Write-DosMessage -Level Error -Message $errorMsg
+        throw $errorMsg
+    }
+}
+
+function Move-ActiveDirectoryGroupsToAzureAD {
+     param(
+        [Parameter(Mandatory=$true)]
+        [string] $connString
+    )
+
+    Write-DosMessage -Level Information -Message "Migrating AD groups to Azure AD..."
+
+    $allowedTenantIds = @()
+
+    try {
+        $allowedTenantIds = Get-AzureADTenants -installConfigPath $installConfigPath
+    }
+    catch {
+        Write-DosMessage -Level Fatal -Message "Error retrieving Azure AD Tenants. Check to ensure the installConfigPath ($($installConfigPath)) is correct."
+        throw
+    }
+    if ($null -eq $allowedTenantIds -or $allowedTenantIds.Count -eq 0) {
+        Write-DosMessage -Level Fatal -Message  "No tenants were found in the install.config"
+        throw
+    }
+
+    # retrieve all groups from Auth DB    
+    try {
+        $results = Get-NonMigratedActiveDirectoryGroups -connectionString $connString
+    }
+    catch {
+        Write-DosMessage -Level Fatal -Message "An error occurred while retrieving non-migrated AD groups. Connection String: $($connString). Error $($_.Exception)"
+        throw
+    }
+
+    if ($results.Count -eq 0) {
+        Write-DosMessage -Level Information -Message "No groups found to migrate."
+    }
+
+    foreach ($group in $results) {
+        # query AD for SID (example SID: S-1-5-21-406681558-3692380417-1824333429-2607)
+        $adGroupSID = ""
+        try {
+            $authGroupNameParts = $group.Name.Split('\')
+            $adGroupName = ""
+            if ($authGroupNameParts.Count -eq 1) {
+                $adGroupName = $authGroupNameParts[0]
```

Comment:
Are there cases where the group name has just a single part for an AD group?

Reply:
I'm not sure :)